### DISCUSSION TITLE
an error while trying to upgrade NetBox from 2.9.11 to 2.11.12

### DISCUSSION THREAD

**Discussion by Jason-MK** [netbox-community/netbox#7185] [General]
Hello , 

I have netox on this path /opt/netbox  without nay number of version beside it but when I open the Web interface the version is 2.9.11 and get this error when I try to upgrade to 2.11.12

`ERROR: Database schema migrations are out of synchronization. (No
data has been lost.) If attempting to upgrade to NetBox v3.0 or
later, first upgrade to a v2.11 release to ensure schema migrations
have been correctly prepared. For further detail on the exact error,
run the following commands:

    source /opt/netbox/venv/bin/activate
    python3 netbox/manage.py showmigrations
--------------------------------------------------------------------
root@NetBox:/opt/netbox# ^C
root@NetBox:/opt/netbox# source /opt/netbox/venv/bin/activate
(venv) root@NetBox:/opt/netbox# ^C
(venv) root@NetBox:/opt/netbox# python3 netbox/manage.py showmigrations
/opt/netbox/netbox/netbox/settings.py:51: UserWarning: The CACHE_TIMEOUT configuration parameter was removed in v3.0.0 and no longer has any                                                                                                   effect.
  "The CACHE_TIMEOUT configuration parameter was removed in v3.0.0 and no longer has any effect."
/opt/netbox/netbox/netbox/settings.py:55: UserWarning: The RELEASE_CHECK_TIMEOUT configuration parameter was removed in v3.0.0 and no longer                                                                                                   has any effect.
  "The RELEASE_CHECK_TIMEOUT configuration parameter was removed in v3.0.0 and no longer has any effect."
Traceback (most recent call last):
  File "/opt/netbox/venv/lib/python3.7/site-packages/django/db/migrations/loader.py", line 259, in build_graph
    self.graph.validate_consistency()
  File "/opt/netbox/venv/lib/python3.7/site-packages/django/db/migrations/graph.py", line 195, in validate_consistency
    [n.raise_error() for n in self.node_map.values() if isinstance(n, DummyNode)]
  File "/opt/netbox/venv/lib/python3.7/site-packages/django/db/migrations/graph.py", line 195, in <listcomp>
    [n.raise_error() for n in self.node_map.values() if isinstance(n, DummyNode)]
  File "/opt/netbox/venv/lib/python3.7/site-packages/django/db/migrations/graph.py", line 58, in raise_error
    raise NodeNotFoundError(self.error_message, self.key, origin=self.origin)
django.db.migrations.exceptions.NodeNotFoundError: Migration dcim.0131_consoleport_speed dependencies reference nonexistent parent node ('dc                                                                                                  im', '0130_sitegroup')

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "netbox/manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/opt/netbox/venv/lib/python3.7/site-packages/django/core/management/__init__.py", line 419, in execute_from_command_line
    utility.execute()
  File "/opt/netbox/venv/lib/python3.7/site-packages/django/core/management/__init__.py", line 413, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/opt/netbox/venv/lib/python3.7/site-packages/django/core/management/base.py", line 354, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/opt/netbox/venv/lib/python3.7/site-packages/django/core/management/base.py", line 398, in execute
    output = self.handle(*args, **options)
  File "/opt/netbox/venv/lib/python3.7/site-packages/django/core/management/commands/showmigrations.py", line 52, in handle
    return self.show_list(connection, options['app_label'])
  File "/opt/netbox/venv/lib/python3.7/site-packages/django/core/management/commands/showmigrations.py", line 71, in show_list
    loader = MigrationLoader(connection, ignore_no_migrations=True)
  File "/opt/netbox/venv/lib/python3.7/site-packages/django/db/migrations/loader.py", line 53, in __init__
    self.build_graph()
  File "/opt/netbox/venv/lib/python3.7/site-packages/django/db/migrations/loader.py", line 284, in build_graph
    ) from exc
django.db.migrations.exceptions.NodeNotFoundError: Migration dcim.0131_consoleport_speed depends on nonexistent node ('dcim', '0130_sitegrou                                                                                                  p'). Django tried to replace migration dcim.0130_sitegroup with any of [dcim.0003_squashed_0130] but wasn't able to because some of the repl                                                                                                  aced migrations are already applied.
`

Is there any way to fix this issue ? I really need your help 

Best Regards

**Comment by jeremystretch** [comment]
> get this error when I try to upgrade to 2.11.12

But:

> UserWarning: The CACHE_TIMEOUT configuration parameter was removed in v3.0.0 and no longer has any effect.

> UserWarning: The RELEASE_CHECK_TIMEOUT configuration parameter was removed in v3.0.0 and no longer has any effect.

These errors indicate that you are trying to upgrade directly to v3.0. You need to upgrade to a v2.11 release before upgrading to v3.0, as called out in the release notes.

**Reply by Jason-MK** [reply]
thanks for the answer 

Which version should I upgrade to ? 

and when I try to do the upgrade for example 

cp /opt/netbox-X.Y.Z/local_requirements.txt /opt/netbox/

which command should I use I have only /opt/netbox with any version number beside it ? 


best regards

**Reply by candlerb** [reply]
> Which version should I upgrade to ?

v2.11.12 first.  When that upgrade is complete and successful (and you *must* have run the `./upgrade.sh` step successfully), go to v3.0.1 - or later if a newer one has been released by then.

> and when I try to do the upgrade for example
> 
> cp /opt/netbox-X.Y.Z/local_requirements.txt /opt/netbox/
> 
> which command should I use I have only /opt/netbox with any version number beside it ?

There are two different ways to install and upgrade Netbox.

Options A: Using release tarballs.  In this case, you untar to /opt/netbox-X.Y.Z and create a symlink from /opt/netbox to /opt/netbox-X.Y.Z.  The instructions for doing this are in the [install instructions](https://netbox.readthedocs.io/en/stable/installation/3-netbox/#option-a-download-a-release-archive) and the [upgrade instructions](https://netbox.readthedocs.io/en/stable/installation/upgrading/#option-a-download-a-release).  Hint: it's the `ln -s` or `ln -sfn` step which creates the symlink.

Option B: Using git.  In this case, you [install](https://netbox.readthedocs.io/en/stable/installation/3-netbox/#option-b-clone-the-git-repository) with a "git checkout", and you [upgrade](https://netbox.readthedocs.io/en/stable/installation/upgrading/#option-b-clone-the-git-repository) using "git pull" on the master branch.  At the moment, the tip of the master branch is v3.0.1

Therefore if you are doing this, then after "git pull" you should do "git checkout v2.11.12" which will temporarily move back to the older version in a "detached" local state (no branch).  Then do the upgrade.  When it's all successful, do "git checkout master" and then upgrade again, and you'll be on v3.0.1

The important thing is: if you originally installed Netbox following Option A then you need to follow the upgrade instructions for Option A.  Ditto for Option B.

**Reply by Jason-MK** [reply]
Thanks a lot ,

I followed the instruction here 

https://netbox.readthedocs.io/en/stable/installation/upgrading/#option-a-download-a-release

But my question I have only netbox without any version numbers beside it ? so when I do upgrade I don't have a version number but on the web interface I can see v2.9.11 ? 

and what should I do if I want to migrate this one to a new one ? 

Best Regards 

sudo cp /opt/netbox-X.Y.Z/netbox/netbox/configuration.py /opt/netbox/netbox/netbox/

**Reply by candlerb** [reply]
Was it you who originally installed Netbox, or someone else?

If it was someone else, you'll have to determine whether they originally used option A or option B.  If it was option A: `/opt/netbox` will be a symlink.  If it was option B: `/opt/netbox` will be a directory, and you'll have a hidden directory `/opt/netbox/.git`.

If you don't know how to determine that, then show the output of the following two commands:

```
ls -ld /opt/netbox
ls -ld /opt/netbox/.git
```

Once you know that, then you'll know the correct upgrade procedure to follow.
